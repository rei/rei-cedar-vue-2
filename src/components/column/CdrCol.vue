<template>
  <div
    v-if="!isRow && !isList"
    class="cdr-col"
    :class="columnClasses"
  >
    <div class="cdr-col__content">
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </div>
  </div>
  <li
    v-else-if="!isRow && isList"
    class="cdr-col"
    :class="columnClasses"
  >
    <div class="cdr-col__content">
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </div>
  </li>
  <li
    v-else-if="isRow && isList"
    class="cdr-row cdr-col"
    :class="columnClasses"
  >
    <cdr-row
      class="cdr-col"
      v-bind="$attrs"
    >
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </cdr-row>
  </li>
  <cdr-row
    v-else
    class="cdr-col"
    :class="columnClasses"
    v-bind="$attrs"
  >
    <!-- @slot innerHTML inside of the column component -->
    <slot/>
  </cdr-row>
</template>

<script>
import { CdrRow } from '@rei/cdr-row';

/**
 * Cedar 2 component for column
 * **Note**: immediate children of `cdr-col` are treated as flex items (due to a flex height bug in Safari). As a result, if you want to have more than one child element they should be wrapped in a single `div` element.
 * **NOTE**: anytime a `cdr-col` is nested within another `cdr-col` the parent needs `is-row` to function correctly
 * See cdr-row for complex example.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrCol',
  components: { CdrRow },
  inject: {
    rowType: { default: 'normal' },
  },
  inheritAttrs: false,
  props: {
    /**
     * Number of columns (1-12) the column should span.
     */
    span: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    spanSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    spanMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    spanLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /**
     * Number of columns (1-12) of empty space to add left of this column.
     */
    offsetLeft: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    offsetLeftSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    offsetLeftMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    offsetLeftLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /**
     * Number of columns (1-12) of empty space to add right of this column.
     */
    offsetRight: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    offsetRightSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    offsetRightMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    offsetRightLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /**
     * How the column should align (overrides cdr-row alignment). Possible values: {top, middle, bottom, stretch}. See CSS flexbox align-self.
     */
    alignSelf: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Sm breakpoint and above */
    alignSelfSm: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Md breakpoint and above */
    alignSelfMd: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Lg breakpoint and above */
    alignSelfLg: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /**
     * Makes the column act as a new cdr-row. Setting this to true exposes the same props as cdr-row.
     */
    isRow: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    isList() {
      return this.rowType === 'list';
    },
    columnClasses() {
      const classObj = {};
      // span
      classObj[`cdr-col_span${this.span}`] = this.span;
      classObj[`cdr-col_span${this.spanSm}@sm`] = this.spanSm;
      classObj[`cdr-col_span${this.spanMd}@md`] = this.spanMd;
      classObj[`cdr-col_span${this.spanLg}@lg`] = this.spanLg;
      // offset left
      classObj[`cdr-col--offsetLeft${this.offsetLeft}`] = this.offsetLeft;
      classObj[`cdr-col--offsetLeft${this.offsetLeftSm}@sm`] = this.offsetLeftSm;
      classObj[`cdr-col--offsetLeft${this.offsetLeftMd}@md`] = this.offsetLeftMd;
      classObj[`cdr-col--offsetLeft${this.offsetLeftLg}@lg`] = this.offsetLeftLg;
      // offset right
      classObj[`cdr-col--offsetRight${this.offsetRight}`] = this.offsetRight;
      classObj[`cdr-col--offsetRight${this.offsetRightSm}@sm`] = this.offsetRightSm;
      classObj[`cdr-col--offsetRight${this.offsetRightMd}@md`] = this.offsetRightMd;
      classObj[`cdr-col--offsetRight${this.offsetRightLg}@lg`] = this.offsetRightLg;
      // align
      classObj[`cdr-col--${this.alignSelf}`] = this.alignSelf;
      classObj[`cdr-col--${this.alignSelfSm}@sm`] = this.alignSelfSm;
      classObj[`cdr-col--${this.alignSelfMd}@md`] = this.alignSelfMd;
      classObj[`cdr-col--${this.alignSelfLg}@lg`] = this.alignSelfLg;

      return classObj;
    },
  },
};
</script>
