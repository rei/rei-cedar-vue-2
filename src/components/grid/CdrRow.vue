<template>
  <component
    :is="type === 'list' ? 'ul' : 'div'"
    :class="[
      $style['cdr-row'],
      colsClass,
      justifyClass,
      alignClass,
      gutterClass,
      verticalClass,
      wrapClass,
      nowrapClass,
    ]"
  >
    <slot />
  </component>
</template>

<script>
import modifier from 'mixinsdir/modifier';
import propValidator from 'srcdir/utils/propValidator';

export default {
  name: 'CdrRow',
  mixins: [modifier],
  props: {
    /**
     * Number of equal-width columns in the row. A value of ‘auto’ will size columns as wide as their content.
     * Possible values: {1-12, auto}
     * Also accepts responsive values with `@breakpoint`: "2 4@md"
     */
    cols: {
      type: String,
      validator: value => propValidator(
        value,
        ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', 'auto'],
      ),
    },
    /**
     * Justify columns within a row. See CSS flexbox justify-content.
     * Possible values: {left, center, right, around, between}.
     * Also accepts responsive values with `@breakpoint`: "center right@lg"
     */
    justify: {
      type: String,
      validator: value => propValidator(
        value,
        ['left', 'center', 'right', 'around', 'between'],
      ),
    },
    /**
     * Align columns of different heights. See CSS flexbox align-items.
     * Possible values: {top, middle, bottom, stretch}.
     * Also accepts responsive values with `@breakpoint`: "top middle@sm"
     */
    align: {
      type: String,
      validator: value => propValidator(
        value,
        ['top', 'middle', 'bottom', 'stretch'],
      ),
    },
    /**
     * Defines gutter size. Default gutter size is 16px @xs and @sm and 32px @md and @lg.
     * Possible values: {none, xxs}.
     * Also accepts responsive values with `@breakpoint`: "none@md"
     */
    gutter: {
      type: String,
      validator: value => propValidator(
        value,
        ['none', 'xxs'],
      ),
    },
    /**
     * Changes row to a column layout. See CSS flexbox flex-direction.
     * Possible values: {vertical}.
     * Also accepts responsive values with `@breakpoint`: "vertical@md"
     */
    vertical: {
      type: String,
      validator: value => propValidator(
        value,
        ['vertical'],
      ),
    },
    /**
     * Enables row wrapping. Only needs to be changed if overriding `nowrap`. See CSS flexbox flex-wrap.
     * Possible values: {wrap}.
     * Also accepts responsive values with `@breakpoint`: "wrap@md"
     */
    wrap: {
      type: String,
      validator: value => propValidator(
        value,
        ['wrap'],
      ),
    },
    /**
     * Disables row wrapping and enables overflow scrolling. See CSS flexbox flex-wrap.
     * Possible values: {nowrap}.
     * Also accepts responsive values with `@breakpoint`: "nowrap@md"
     */
    nowrap: {
      type: String,
      validator: value => propValidator(
        value,
        ['nowrap'],
      ),
    },
    /**
     * Sets grid to use `ul` and `li` or `div`.
     * Possible values: {normal, list}.
     */
    type: {
      type: String,
      default: 'normal',
      validator: value => propValidator(
        value,
        ['normal', 'list'],
        false,
      ),
    },
  },
  provide() {
    return {
      rowType: this.type,
    };
  },
  computed: {
    colsClass() {
      const classStr = [];

      if (this.cols) {
        this.cols.split(' ').forEach((val) => {
          classStr.push(this.$style[`cdr-row_row${val}`]);
        });
      }

      return classStr.join(' ');
    },
    justifyClass() {
      const classStr = [];

      if (this.justify) {
        this.justify.split(' ').forEach((val) => {
          classStr.push(this.modifyClassName('cdr-row', val));
        });
      }

      return classStr.join(' ');
    },
    alignClass() {
      const classStr = [];

      if (this.align) {
        this.align.split(' ').forEach((val) => {
          classStr.push(this.modifyClassName('cdr-row', val));
        });
      }

      return classStr.join(' ');
    },
    gutterClass() {
      const classStr = [];

      if (this.gutter) {
        this.gutter.split(' ').forEach((val) => {
          classStr.push(this.$style[`cdr-row--gutter-${val}`]);
        });
      }

      return classStr.join(' ');
    },
    verticalClass() {
      const classStr = [];

      if (this.vertical) {
        this.vertical.split(' ').forEach((val) => {
          classStr.push(this.modifyClassName('cdr-row', val));
        });
      }

      return classStr.join(' ');
    },
    wrapClass() {
      const classStr = [];

      if (this.wrap) {
        this.wrap.split(' ').forEach((val) => {
          classStr.push(this.modifyClassName('cdr-row', val));
        });
      }

      return classStr.join(' ');
    },
    nowrapClass() {
      const classStr = [];

      if (this.nowrap) {
        this.nowrap.split(' ').forEach((val) => {
          classStr.push(this.modifyClassName('cdr-row', val));
        });
      }

      return classStr.join(' ');
    },
  },
};
</script>

<style module>
@import 'cssdir/settings/_index.pcss';
@import "./styles/Grid.vars.pcss";
@import "./styles/CdrRow.pcss";
</style>
