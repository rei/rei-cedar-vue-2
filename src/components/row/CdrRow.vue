<template>
  <div
    v-if="type === 'normal'"
    class="cdr-row"
    :class="rowClasses"
  >
    <!-- @slot innerHTML inside the row component -->
    <slot/>
  </div>
  <ul
    v-else-if="type === 'list'"
    class="cdr-row"
    :class="rowClasses"
  >
    <!-- @slot innerHTML inside the row component -->
    <slot/>
  </ul>
</template>

<script>
/**
 * Cedar 2 component for row
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrRow',
  props: {
    /**
     * Number of columns in the row (1-12, auto).
     */
    cols: {
      type: [String, Number],
      validator: (value) => {
        const str = value.toString();
        return ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', 'auto']
          .indexOf(str) >= 0 || false;
      },
    },
    /** Sm breakpoint and above */
    colsSm: {
      type: [String, Number],
      validator: (value) => {
        const str = value.toString();
        return ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', 'auto']
          .indexOf(str) >= 0 || false;
      },
    },
    /** Md breakpoint and above */
    colsMd: {
      type: [String, Number],
      validator: (value) => {
        const str = value.toString();
        return ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', 'auto']
          .indexOf(str) >= 0 || false;
      },
    },
    /** Lg breakpoint and above */
    colsLg: {
      type: [String, Number],
      validator: (value) => {
        const str = value.toString();
        return ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', 'auto']
          .indexOf(str) >= 0 || false;
      },
    },
    /**
     * How columns should be justified within empty space of the row. Possible values: {left, center, right, around, between}. See CSS flexbox justify-content.
     */
    justify: {
      type: String,
      validator: value => (['left', 'center', 'right', 'around', 'between']
        .indexOf(value) >= 0) || false,
    },
    /** Sm breakpoint and above */
    justifySm: {
      type: String,
      validator: value => (['left', 'center', 'right', 'around', 'between']
        .indexOf(value) >= 0) || false,
    },
    /** Md breakpoint and above */
    justifyMd: {
      type: String,
      validator: value => (['left', 'center', 'right', 'around', 'between']
        .indexOf(value) >= 0) || false,
    },
    /** Lg breakpoint and above */
    justifyLg: {
      type: String,
      validator: value => (['left', 'center', 'right', 'around', 'between']
        .indexOf(value) >= 0) || false,
    },
    /**
     * How columns of different heights should align. {top, middle, bottom, stretch}. See CSS flexbox align-items.
     */
    align: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Sm breakpoint and above */
    alignSm: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Md breakpoint and above */
    alignMd: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Lg breakpoint and above */
    alignLg: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /**
     * Overrides default gutter spacing. {none}.
     */
    gutter: {
      type: String,
      validator: value => (['none', 'xxs'].indexOf(value) >= 0) || false,
    },
    /** Sm breakpoint and above */
    gutterSm: {
      type: String,
      validator: value => (['none', 'xxs'].indexOf(value) >= 0) || false,
    },
    /** Md breakpoint and above */
    gutterMd: {
      type: String,
      validator: value => (['none', 'xxs'].indexOf(value) >= 0) || false,
    },
    /** Lg breakpoint and above */
    gutterLg: {
      type: String,
      validator: value => (['none', 'xxs'].indexOf(value) >= 0) || false,
    },
    /**
     * Changes row to a column layout. See CSS flexbox flex-direction.
     */
    vertical: {
      type: Boolean,
      default: false,
    },
    /** Sm breakpoint and above */
    verticalSm: {
      type: Boolean,
      default: false,
    },
    /** Md breakpoint and above */
    verticalMd: {
      type: Boolean,
      default: false,
    },
    /** Lg breakpoint and above */
    verticalLg: {
      type: Boolean,
      default: false,
    },
    /**
     * Enables row wrapping. Only needs to be changed if overriding `nowrap`. See CSS flexbox flex-wrap.
     */
    wrap: {
      type: Boolean,
      default: false,
    },
    /** Sm breakpoint and above */
    wrapSm: {
      type: Boolean,
      default: false,
    },
    /** Md breakpoint and above */
    wrapMd: {
      type: Boolean,
      default: false,
    },
    /** Lg breakpoint and above */
    wrapLg: {
      type: Boolean,
      default: false,
    },
    /**
     * Disables row wrapping and enables overflow scrolling. See CSS flexbox flex-wrap.
     */
    nowrap: {
      type: Boolean,
      default: false,
    },
    /** Sm breakpoint and above */
    nowrapSm: {
      type: Boolean,
      default: false,
    },
    /** Md breakpoint and above */
    nowrapMd: {
      type: Boolean,
      default: false,
    },
    /** Lg breakpoint and above */
    nowrapLg: {
      type: Boolean,
      default: false,
    },
    /**
     * Changes grid to built with ul>li rather than divs. {normal, list}
     */
    type: {
      type: String,
      default: 'normal',
      validator: value => (['normal', 'list'].indexOf(value) >= 0) || false,
    },
  },
  provide() {
    return {
      rowType: this.type,
    };
  },
  computed: {
    rowClasses() {
      const classObj = {};
      classObj[`cdr-row_row${this.cols}`] = this.cols;
      classObj[`cdr-row_row${this.colsSm}@sm`] = this.colsSm;
      classObj[`cdr-row_row${this.colsMd}@md`] = this.colsMd;
      classObj[`cdr-row_row${this.colsLg}@lg`] = this.colsLg;
      // justify
      classObj[`cdr-row--${this.justify}`] = this.justify;
      classObj[`cdr-row--${this.justifySm}@sm`] = this.justifySm;
      classObj[`cdr-row--${this.justifyMd}@md`] = this.justifyMd;
      classObj[`cdr-row--${this.justifyLg}@lg`] = this.justifyLg;
      // align
      classObj[`cdr-row--${this.align}`] = this.align;
      classObj[`cdr-row--${this.alignSm}@sm`] = this.alignSm;
      classObj[`cdr-row--${this.alignMd}@md`] = this.alignMd;
      classObj[`cdr-row--${this.alignLg}@lg`] = this.alignLg;
      // column
      classObj['cdr-row--column'] = this.vertical;
      classObj['cdr-row--column@sm'] = this.verticalSm;
      classObj['cdr-row--column@md'] = this.verticalMd;
      classObj['cdr-row--column@lg'] = this.verticalLg;
      // gutter
      classObj[`cdr-row--gutter-${this.gutter}`] = this.gutter;
      classObj[`cdr-row--gutter-${this.gutterSm}@sm`] = this.gutterSm;
      classObj[`cdr-row--gutter-${this.gutterMd}@md`] = this.gutterMd;
      classObj[`cdr-row--gutter-${this.gutterLg}@lg`] = this.gutterLg;
      // wrap
      classObj['cdr-row--wrap@sm'] = this.wrapSm;
      classObj['cdr-row--wrap@md'] = this.wrapMd;
      classObj['cdr-row--wrap@lg'] = this.wrapLg;
      // noWrap
      classObj['cdr-row--noWrap'] = this.nowrap;
      classObj['cdr-row--noWrap@sm'] = this.nowrapSm;
      classObj['cdr-row--noWrap@md'] = this.nowrapMd;
      classObj['cdr-row--noWrap@lg'] = this.nowrapLg;

      return classObj;
    },
  },
};
</script>
