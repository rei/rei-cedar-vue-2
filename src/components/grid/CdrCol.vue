<template>
  <!-- !row && !list -->
  <div
    v-if="!isRow && !isList"
    :class="[$style['cdr-col'],columnClasses]"
  >
    <div :class="$style['cdr-col__content']">
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </div>
  </div>

  <!-- !row && list -->
  <li
    v-else-if="!isRow && isList"
    :class="[$style['cdr-col'],columnClasses]"
  >
    <div :class="$style['cdr-col__content']">
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </div>
  </li>

  <!-- row && list -->
  <li
    v-else-if="isRow && isList"
    :class="[$style['cdr-row'], $style['cdr-col'],columnClasses]"
  >
    <cdr-row
      :class="$style['cdr-col']"
      v-bind="$attrs"
    >
      <!-- @slot innerHTML inside of the column component -->
      <slot/>
    </cdr-row>
  </li>

  <!-- else -->
  <cdr-row
    v-else
    :class="[$style['cdr-col'],columnClasses]"
    v-bind="$attrs"
  >
    <!-- @slot innerHTML inside of the column component -->
    <slot/>
  </cdr-row>
</template>

<script>
import CdrRow from './CdrRow';

/**
 * Cedar 2 component for column
 * **Note**: immediate children of `cdr-col` are treated as flex items (due to a flex height bug in Safari). As a result, if you want to have more than one child element they should be wrapped in a single `div` element.
 * **NOTE**: anytime a `cdr-col` is nested within another `cdr-col` the parent needs `is-row` to function correctly
 * See cdr-row for complex example.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrCol',
  components: { CdrRow },
  inject: {
    rowType: { default: 'normal' },
  },
  inheritAttrs: false,
  props: {
    /**
     * Number of columns (1-12) the column should span.
     */
    span: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    spanSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    spanMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    spanLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num > 0 && num <= 12) || false;
      },
    },
    /**
     * Number of columns (1-12) of empty space to add left of this column.
     */
    offsetLeft: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    offsetLeftSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    offsetLeftMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    offsetLeftLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /**
     * Number of columns (1-12) of empty space to add right of this column.
     */
    offsetRight: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Sm breakpoint and above */
    offsetRightSm: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Md breakpoint and above */
    offsetRightMd: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /** Lg breakpoint and above */
    offsetRightLg: {
      type: [String, Number],
      validator: (value) => {
        const num = parseInt(value, 10);
        return (num >= 0 && num <= 12) || false;
      },
    },
    /**
     * How the column should align (overrides cdr-row alignment). Possible values: {top, middle, bottom, stretch}. See CSS flexbox align-self.
     */
    alignSelf: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Sm breakpoint and above */
    alignSelfSm: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Md breakpoint and above */
    alignSelfMd: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /** Lg breakpoint and above */
    alignSelfLg: {
      type: String,
      validator: value => (['top', 'middle', 'bottom', 'stretch'].indexOf(value) >= 0) || false,
    },
    /**
     * Makes the column act as a new cdr-row. Setting this to true exposes the same props as cdr-row.
     */
    isRow: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    isList() {
      return this.rowType === 'list';
    },
    columnClasses() {
      const classObj = {};
      // span
      classObj[this.$style[`cdr-col_span${this.span}`]] = this.span;
      classObj[this.$style[`cdr-col_span${this.spanSm}@sm`]] = this.spanSm;
      classObj[this.$style[`cdr-col_span${this.spanMd}@md`]] = this.spanMd;
      classObj[this.$style[`cdr-col_span${this.spanLg}@lg`]] = this.spanLg;
      // offset left
      classObj[this.$style[`cdr-col--offsetLeft${this.offsetLeft}`]] = this.offsetLeft;
      classObj[this.$style[`cdr-col--offsetLeft${this.offsetLeftSm}@sm`]] = this.offsetLeftSm;
      classObj[this.$style[`cdr-col--offsetLeft${this.offsetLeftMd}@md`]] = this.offsetLeftMd;
      classObj[this.$style[`cdr-col--offsetLeft${this.offsetLeftLg}@lg`]] = this.offsetLeftLg;
      // offset right
      classObj[this.$style[`cdr-col--offsetRight${this.offsetRight}`]] = this.offsetRight;
      classObj[this.$style[`cdr-col--offsetRight${this.offsetRightSm}@sm`]] = this.offsetRightSm;
      classObj[this.$style[`cdr-col--offsetRight${this.offsetRightMd}@md`]] = this.offsetRightMd;
      classObj[this.$style[`cdr-col--offsetRight${this.offsetRightLg}@lg`]] = this.offsetRightLg;
      // align
      classObj[this.$style[`cdr-col--${this.alignSelf}`]] = this.alignSelf;
      classObj[this.$style[`cdr-col--${this.alignSelfSm}@sm`]] = this.alignSelfSm;
      classObj[this.$style[`cdr-col--${this.alignSelfMd}@md`]] = this.alignSelfMd;
      classObj[this.$style[`cdr-col--${this.alignSelfLg}@lg`]] = this.alignSelfLg;

      return classObj;
    },
  },
};
</script>

<style module>
@import 'cssdir/settings/_index.pcss';
@import "./styles/Grid.vars.pcss";
@import "./styles/CdrCol.pcss";
</style>
