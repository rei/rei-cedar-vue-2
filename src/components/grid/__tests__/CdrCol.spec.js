import { shallowMount, mount } from '@vue/test-utils';
import CdrCol from 'componentdir/grid/CdrCol';
import CdrRow from 'componentdir/grid/CdrRow';

describe('CdrCol', () => {
  it('renders as list type correctly', () => {
    const wrapper = mount(CdrRow, {
      propsData: {
        type: 'list',
      },
      slots: {
        default: CdrCol
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders as list type correctly when isRow', () => {
    const wrapper = mount(CdrRow, {
      propsData: {
        type: 'list',
      },
      stubs: {
        'cdr-col': CdrCol,
      },
      slots: {
        default: ['<cdr-col is-row>column</cdr-col>']
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders default correctly', () => {
    const wrapper = mount(CdrCol);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has the correct block class', () => {
    const wrapper = shallowMount(CdrCol);
    expect(wrapper.classes()).toContain('cdr-col');
  });

  it('computes span classes correctly', () => {
    const wrapper = shallowMount(CdrCol, {
      propsData: {
        span: '1 2@sm 2@md 2@lg',
      }
    });
    expect(wrapper.classes()).toContain('cdr-col_span1');
    expect(wrapper.classes()).toContain('cdr-col_span2@md');
    expect(wrapper.classes()).toContain('cdr-col_span2@lg');
    expect(wrapper.classes()).toContain('cdr-col_span2@sm');
  });

  it('computes offsetLeft classes correctly', () => {
    const wrapper = shallowMount(CdrCol, {
      propsData: {
        offsetLeft: '1 2@sm 2@md 2@lg',
      }
    });
    expect(wrapper.classes()).toContain('cdr-col--offsetLeft1');
    expect(wrapper.classes()).toContain('cdr-col--offsetLeft2@md');
    expect(wrapper.classes()).toContain('cdr-col--offsetLeft2@lg');
    expect(wrapper.classes()).toContain('cdr-col--offsetLeft2@sm');
  });

  it('computes offsetRight classes correctly', () => {
    const wrapper = shallowMount(CdrCol, {
      propsData: {
        offsetRight: '1 2@sm 2@md 2@lg',
      }
    });
    expect(wrapper.classes()).toContain('cdr-col--offsetRight1');
    expect(wrapper.classes()).toContain('cdr-col--offsetRight2@md');
    expect(wrapper.classes()).toContain('cdr-col--offsetRight2@lg');
    expect(wrapper.classes()).toContain('cdr-col--offsetRight2@sm');
  });

  it('computes alignSelf classes correctly', () => {
    const wrapper = shallowMount(CdrCol, {
      propsData: {
        alignSelf: 'top top@sm top@md top@lg',
      }
    });
    expect(wrapper.classes()).toContain('cdr-col--top');
    expect(wrapper.classes()).toContain('cdr-col--top@md');
    expect(wrapper.classes()).toContain('cdr-col--top@lg');
    expect(wrapper.classes()).toContain('cdr-col--top@sm');
  });

  it('computes isRow class correctly', () => {
    const wrapper = mount(CdrCol, {
      propsData: {
        isRow: true,
      }
    });
    expect(wrapper.classes()).toContain('cdr-row');
  });
});