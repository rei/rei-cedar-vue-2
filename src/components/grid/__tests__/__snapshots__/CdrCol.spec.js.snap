// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrCol renders as list type correctly 1`] = `
<ul
  class="cdr-row"
>
  <li
    class="cdr-col"
  >
    <div
      class="cdr-col__content"
    />
  </li>
</ul>
`;

exports[`CdrCol renders as list type correctly when isRow 1`] = `
<ul
  class="cdr-row"
>
  <li
    class="cdr-row cdr-col"
  >
    <div
      class="cdr-row cdr-col"
    >
      column
    </div>
  </li>
</ul>
`;

exports[`CdrCol renders default correctly 1`] = `
<div
  class="cdr-col"
>
  <div
    class="cdr-col__content"
  />
</div>
`;
